internal/apiclient/policyv2.go

```diff
@@ -0,0 +1,294 @@
+package apiclient
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"github.com/infracost/infracost/internal/config"
+	"github.com/infracost/infracost/internal/schema"
+	"sort"
+)
+
+type PolicyV2APIClient struct {
+	APIClient
+	allowlist resourceAllowlist
+}
+
+// NewPolicyV2APIClient retrieves resource allowlist info from Infracost Cloud and returns a new policy client
+func NewPolicyV2APIClient(ctx *config.RunContext) (*PolicyV2APIClient, error) {
+	c := PolicyV2APIClient{
+		APIClient: APIClient{
+			endpoint: ctx.Config.TagPolicyAPIEndpoint, // use tag policy api endpoint for now.
+			apiKey:   ctx.Config.APIKey,
+			uuid:     ctx.UUID(),
+		},
+	}
+
+	prw, err := c.getPolicyResourceAllowlist()
+	if err != nil {
+		return nil, err
+	}
+	c.allowlist = prw
+
+	return &c, nil
+}
+
+// UploadPolicyData sends a filtered set of a project's resource information to Infracost Cloud and
+// potentially adds PolicySha and PastPolicySha to the project's metadata.
+func (c *PolicyV2APIClient) UploadPolicyData(project *schema.Project) error {
+	if project.Metadata == nil {
+		project.Metadata = &schema.ProjectMetadata{}
+	}
+
+	filteredResources := c.filterResources(project.PartialResources)
+	if len(filteredResources) > 0 {
+		sha, err := c.uploadProjectPolicyData(filteredResources)
+		if err != nil {
+			return fmt.Errorf("failed to upload filtered partial resources %w", err)
+		}
+		project.Metadata.PolicySha = sha
+	}
+
+	filteredPastResources := c.filterResources(project.PartialPastResources)
+	if len(filteredPastResources) > 0 {
+		sha, err := c.uploadProjectPolicyData(filteredPastResources)
+		if err != nil {
+			return fmt.Errorf("failed to upload filtered past partial resources %w", err)
+		}
+		project.Metadata.PastPolicySha = sha
+	}
+
+	return nil
+}
+
+func (c *PolicyV2APIClient) uploadProjectPolicyData(p2rs []policy2Resource) (string, error) {
```

Comment:
do we want to add some retry logic to this - or is it already handled in the base client?